cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java

```diff
@@ -182,6 +184,54 @@ public void recordProgramStop(Id.Program program, String pid, long stopTs, Progr
     }
   }
 
+  public RunRecord getRun(Id.Program program, final String runid) {
+    // Query running records first
+    MDSKey runningKey = new MDSKey.Builder()
+      .add(TYPE_RUN_RECORD_STARTED)
+      .add(program.getNamespaceId())
+      .add(program.getApplicationId())
+      .add(program.getType().name())
+      .add(program.getId())
+      .add(runid)
+      .build();
+    RunRecord running = get(runningKey, RunRecord.class);
+    // If program is running, this will be non-null
+    if (running != null) {
+      return running;
+    }
+
+    // If program is not running, query completed run records
+    MDSKey completedKey = new MDSKey.Builder()
+      .add(TYPE_RUN_RECORD_COMPLETED)
+      .add(program.getNamespaceId())
+      .add(program.getApplicationId())
+      .add(program.getType().name())
+      .add(program.getId())
+      .build();
+
+    // Get start time from RunId
+    long programStartMillis = RunIds.getTimeMillis(RunIds.fromString(runid));
+    if (programStartMillis > -1) {
+      // If start time is found, run a get
+      long programStartSecs = TimeUnit.MILLISECONDS.toSeconds(programStartMillis);
+      return get(new MDSKey.Builder(completedKey).add(getInvertedTsKeyPart(programStartSecs)).add(runid).build(),
```

Comment:
`getInvertedTsKeyPart(programStartSecs)` is part of the key for completed run record.